singularity-eos/eos/eos_stiff.hpp

```diff
@@ -0,0 +1,200 @@
+  }
+  StiffGas GetOnDevice() { return *this; }
+  PORTABLE_INLINE_FUNCTION Real TemperatureFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const {
+    return MYMAX( 0.0, (rho*(sie-_qq)-_Pinf) / (rho*_Cv) );
+  }
+  PORTABLE_INLINE_FUNCTION void checkParams() const {
+    // Portable_require seems to do the opposite of what it should. Conditions
+    // reflect this and the code should be changed when ports-of-call changes
+    PORTABLE_ALWAYS_REQUIRE(_Cv >= 0, "Heat capacity must be positive");
+    PORTABLE_ALWAYS_REQUIRE(_gm1 >= 0, "Gruneisen parameter must be positive");
+  }
+  PORTABLE_INLINE_FUNCTION Real InternalEnergyFromDensityTemperature(
+      const Real rho, const Real temperature, Real *lambda = nullptr) const {
+    return MYMAX( 0.0, (rho*_Cv*temperature+_Pinf)/rho + _qq );
+  }
+  PORTABLE_INLINE_FUNCTION Real PressureFromDensityTemperature(
+      const Real rho, const Real temperature, Real *lambda = nullptr) const {
+    return MYMAX(0.0, _gm1 * rho * _Cv * temperature - _Pinf );
+  }
+  PORTABLE_INLINE_FUNCTION Real PressureFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const {
+    return MYMAX(0.0, _gm1 * rho * (sie-_qq) - (_gm1+1.0)*_Pinf );
+  }
+  PORTABLE_INLINE_FUNCTION Real EntropyFromDensityTemperature(
+      const Real rho, const Real temperature, Real *lambda = nullptr) const {
+    return _Cv * log( temperature / (pow(rho*_gm1*_Cv,_gm1)) ) + _qp;
+  }
+  PORTABLE_INLINE_FUNCTION Real EntropyFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const {
+    //const Real temp = TemperatureFromDensityInternalEnergy(rho, sie, lambda);
+    //return EntropyFromDensityTemperature(rho, temp, lambda);
+    return _Cv * log( (rho*(sie-_qq)-_Pinf) /
+           (pow(_Cv*rho,_gm1+1.0)*pow(_gm1,_gm1)) ) +
+           _qp;
+  }
+  PORTABLE_INLINE_FUNCTION Real SpecificHeatFromDensityTemperature(
+      const Real rho, const Real temperature, Real *lambda = nullptr) const {
+    return _Cv;
+  }
+  PORTABLE_INLINE_FUNCTION Real SpecificHeatFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const {
+    return _Cv;
+  }
+  PORTABLE_INLINE_FUNCTION Real BulkModulusFromDensityTemperature(
+      const Real rho, const Real temperature, Real *lambda = nullptr) const {
+    const Real entr = EntropyFromDensityTemperature(rho, temperature, lambda);
+    return MYMAX(0.0, pow(rho ,_gm1    ) * 
+                      pow(_Cv ,_gm1+1.0) * 
+                      pow(_gm1,_gm1+2.0) * exp((entr-_qp)/_Cv) );
+  }
+  PORTABLE_INLINE_FUNCTION Real BulkModulusFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const {
+    const Real entr = EntropyFromDensityInternalEnergy(rho, sie, lambda);
+    return MYMAX(0.0, pow(rho ,_gm1    ) * 
+                      pow(_Cv ,_gm1+1.0) * 
+                      pow(_gm1,_gm1+2.0) * exp((entr-_qp)/_Cv) );
```

Comment:
I'm worried about the numerics of the exponent and possible overflow/underflow. Can you derive a formula for the bulk modulus from the more general derivatives like we talked about? I.e.
```math
B_S = \rho \left(\frac{\partial P}{\partial \rho}\right)_e + \frac{P}{\rho} \left(\frac{\partial P}{\partial e}\right)_\rho 
```

Reply:
Bulk modulus equation changed to the suggested